tests/beta/tapes/test_tape.py

```diff
@@ -0,0 +1,667 @@
+
+        # check that operations are inverted
+        assert ops[0].inverse
+        assert not ops[1].inverse
+        assert ops[2].inverse
+
+        # check that parameter order has reversed
+        assert tape.get_parameters() == [init_state, p[1], p[2], p[3], p[0]]
+
+    def test_parameter_transforms(self):
+        """Test that inversion correctly changes trainable parameters"""
+        init_state = np.array([1, 1])
+        p = [0.1, 0.2, 0.3, 0.4]
+
+        with QuantumTape() as tape:
+            prep = qml.BasisState(init_state, wires=[0, "a"])
+            ops = [qml.RX(p[0], wires=0), qml.Rot(*p[1:], wires=0).inv(), qml.CNOT(wires=[0, "a"])]
+            m1 = probs(wires=0)
+            m2 = probs(wires="a")
+
+        tape.trainable_params = {1, 2}
+        tape.inv()
+
+        # check that operation order is reversed
+        assert tape.trainable_params == {1, 4}
+        assert tape.get_parameters() == [p[1], p[0]]
+
+        # undo the inverse
+        tape.inv()
+        assert tape.trainable_params == {1, 2}
+        assert tape.get_parameters() == [p[0], p[1]]
+        assert tape._ops == ops
+
+
+class TestExpand:
+    """Tests for tape expansion"""
+
+    def test_decomposition(self):
+        """Test expanding a tape with operations that have decompositions"""
+        with QuantumTape() as tape:
+            qml.Rot(0.1, 0.2, 0.3, wires=0)
+
+        new_tape = tape.expand()
+
+        assert len(new_tape.operations) == 3
+        assert new_tape.get_parameters() == [0.1, 0.2, 0.3]
+        assert new_tape.trainable_params == {0, 1, 2}
+
+        assert isinstance(new_tape.operations[0], qml.RZ)
+        assert isinstance(new_tape.operations[1], qml.RY)
+        assert isinstance(new_tape.operations[2], qml.RZ)
+
+        # check that modifying the new tape does not affect the old tape
+
+        new_tape.trainable_params = {0}
+        new_tape.set_parameters([10])
+
+        assert tape.get_parameters() == [0.1, 0.2, 0.3]
+        assert tape.trainable_params == {0, 1, 2}
+
+    def test_decomposition_removing_parameters(self):
+        """Test that decompositions which reduce the number of parameters
+        on the tape retain tape consistency."""
+        with QuantumTape() as tape:
+            qml.BasisState(np.array([1]), wires=0)
+
+        new_tape = tape.expand()
+
+        assert len(new_tape.operations) == 1
+        assert new_tape.num_params == 0
+        assert new_tape.get_parameters() == []
+
+        assert isinstance(new_tape.operations[0], qml.PauliX)
+
+    def test_decomposition_adding_parameters(self):
+        """Test that decompositions which increase the number of parameters
+        on the tape retain tape consistency."""
+        with QuantumTape() as tape:
+            qml.PauliX(wires=0)
+
+        new_tape = tape.expand()
+
+        assert len(new_tape.operations) == 3
+        assert new_tape.num_params == 3
+        assert new_tape.get_parameters() == [np.pi / 2, np.pi, np.pi / 2]
```

Comment:
Why not check the operations as well? (same for above test)

Reply:
done